src/bt_peer_connection.cpp

```diff
@@ -121,6 +121,35 @@ namespace {
 } // anonymous namespace
 #endif
 
+bool validate_hash_request(hash_request const& hr, file_storage const& fs)
+{
+	if (hr.file < file_index_t{0}
+		|| hr.file >= fs.end_file()
+		|| hr.base < 0
+		|| hr.index < 0
+		|| hr.count <= 0
+		|| hr.count > 8192
+		|| hr.proof_layers < 0)
+		return false;
+
+	int const num_leafs = merkle_num_leafs(fs.file_num_pieces(hr.file));
+	int const num_layers = merkle_num_layers(num_leafs);
+
+	if (hr.base >= num_layers) return false;
+
+	// the number of hashes at the specified level
+	int const level_size = num_leafs / (1 << hr.base);
+
+	// [index, index + count] must be within the number of nodes at the specified
+	// level
+	if (hr.index >= level_size || hr.index + hr.count > level_size)
+		return false;
+
+	if (hr.proof_layers > num_layers - hr.base) return false;
```

Comment:
The comparison should be `>=` because `proof_layers` doesn't include the base layer and is not allowed to include the root node.

Reply:
ok. I'm not entirely sure what `proof_layers` is. It looks like it's the number of "uncle hashes" to send a long. Is that right? If so, it could be implied rather than explicitly requested, no?

Reply:
I noticed that in your calculations you do `merkle_num_layers(merkle_num_leafs(req.count)) - 1`, but I suppose that's the same thing as `merkle_num_layers(merkle_num_leafs(fs.file_num_blocks(hr.file))) - req.base - 1` (as I do in the validation function), right?

Reply:
The idea with `proof_layers` is to let the requester pick how many, if any, uncle hashes it wants to receive. This allows for different hash request strategies, like requesting the piece layer with no uncle hashes and verifying the whole layer once it's completely downloaded.

Reply:
> I noticed that in your calculations you do `merkle_num_layers(merkle_num_leafs(req.count)) - 1`, but I suppose that's the same thing as `merkle_num_layers(merkle_num_leafs(fs.file_num_blocks(hr.file))) - req.base - 1` (as I do in the validation function), right?

Yes, it's the same concept, just in slightly different contexts. The result of the first snippet is the number of proof layers which can be computed using just the requested base layer hashes (i.e. the number of proof layers for which we don't need to send uncle hashes), while the second is the maximum number of proof layers for a given file and base layer.